### PR TITLE
cumprod_superset

### PR DESCRIPTION
**Pull request by Ishticode**
Additional argument `reverse` based on [tf cumprod](https://www.tensorflow.org/api_docs/python/tf/math/cumprod). Test passing.  Logic similar to ivy.cumsum for all backends.
will add some docstring examples.

### CODE REVIEW

File: ivy/functional/backends/jax/statistical.py [hunk 1]
```diff
@@ -142,21 +142,34 @@ def cumprod(
...
-        x = jnp.swapaxes(x, axis, -1)
-        x = jnp.concatenate((jnp.ones_like(x[..., -1:]), x[..., :-1]), -1)
-        res = jnp.cumprod(x, -1)
-        return jnp.swapaxes(res, axis, -1)
-    return jnp.cumprod(x, axis)
+        if dtype is jnp.bool_:
+            dtype = ivy.default_int_dtype(as_native=True)
+        else:
+            dtype = _infer_dtype(x.dtype)
+    if exclusive or reverse:
```
Comment (djl11) [inline ivy/functional/backends/jax/statistical.py:156]: no need for `if exclusive or reverse:`

File: ivy/functional/backends/jax/statistical.py [hunk 2]
```diff
@@ -142,21 +142,34 @@ def cumprod(
...
+        if exclusive and reverse:
+            x = jnp.cumprod(jnp.flip(x, axis=(axis,)), axis=axis, dtype=dtype)
+            x = jnp.swapaxes(x, axis, -1)
+            x = jnp.concatenate((jnp.zeros_like(x[..., -1:]), x[..., :-1]), -1)
+            x = jnp.swapaxes(x, axis, -1)
+            res = jnp.flip(x, axis=(axis,))
```
Comment (djl11) [inline ivy/functional/backends/jax/statistical.py:162]: `return jnp.flip(x, axis=(axis,))`

File: ivy/functional/backends/jax/statistical.py [hunk 3]
```diff
@@ -142,21 +142,34 @@ def cumprod(
...
+        elif exclusive:
+            x = jnp.swapaxes(x, axis, -1)
+            x = jnp.concatenate((jnp.zeros_like(x[..., -1:]), x[..., :-1]), -1)
+            x = jnp.cumprod(x, -1, dtype=dtype)
+            res = jnp.swapaxes(x, axis, -1)
```
Comment (djl11) [inline ivy/functional/backends/jax/statistical.py:167]: `return jnp.swapaxes(x, axis, -1)`

File: ivy/functional/backends/jax/statistical.py [hunk 4]
```diff
@@ -142,21 +142,34 @@ def cumprod(
...
+        elif reverse:
+            x = jnp.cumprod(jnp.flip(x, axis=(axis,)), axis=axis, dtype=dtype)
+            res = jnp.flip(x, axis=axis)
```
Comment (djl11) [inline ivy/functional/backends/jax/statistical.py:170]: `return jnp.flip(x, axis=axis)`

File: ivy/functional/backends/jax/statistical.py [hunk 5]
```diff
@@ -142,21 +142,34 @@ def cumprod(
...
+        return res
+    return jnp.cumprod(x, axis, dtype=dtype)
```
Comment (djl11) [inline ivy/functional/backends/jax/statistical.py:172]: if `exclusive=False` and `reverse=False` is the most comment setup, then put this at the beginning of the conditional block under:

```
if not (exclusive or reverse):
    return jnp.cumprod(x, axis, dtype=dtype)
elif exclusive:
    ...
    return ...
elif reverse:
    ...
    return ...
else:
    ...
    return ...
```

This will avoid repeated boolean checks for the most common operating mode.

File: ivy/functional/backends/numpy/statistical.py [hunk 6]
```diff
@@ -163,18 +163,30 @@ def cumprod(
...
-        x = np.concatenate((np.ones_like(x[..., -1:]), x[..., :-1]), -1)
-        res = np.cumprod(x, -1, dtype=dtype)
-        res = np.swapaxes(res, axis, -1)
-        if out is not None:
-            return ivy.inplace_update(out, res)
+        if x.dtype == "bool":
+            dtype = ivy.default_int_dtype(as_native=True)
+        else:
+            dtype = _infer_dtype(x.dtype)
+    if exclusive or reverse:
```
Comment (djl11) [inline ivy/functional/backends/numpy/statistical.py:175]: same comments as above.

File: ivy/functional/backends/torch/statistical.py [hunk 7]
```diff
@@ -210,21 +210,30 @@ def cumprod(
...
-            dtype = ivy.default_int_dtype(as_native=True)
-        else:
-            dtype = _infer_dtype(x.dtype)
-    if exclusive:
-        x = torch.transpose(x, axis, -1)
-        x = torch.cat((torch.ones_like(x[..., -1:]), x[..., :-1]), -1, out=out)
-        res = torch.cumprod(x, -1, dtype=dtype, out=out)
-        return torch.transpose(res, axis, -1)
+        dtype = _infer_dtype(x.dtype)
+    if exclusive or reverse:
```
Comment (djl11) [inline ivy/functional/backends/torch/statistical.py:221]: same comments as above.

**Review by djl11 — CHANGES_REQUESTED** [review]

**Comment by djl11** [comment]
Once these small changes are made, I think it's good to merge provided that the tests are passing!

**Comment by Ishticode** [comment]
updated and tests passing
